lib/datadog/core/environment/vm_cache.rb

```diff
@@ -0,0 +1,46 @@
+# typed: true
+module Datadog
+  module Core
+    module Environment
+      # Reports Ruby VM cache performance statistics.
+      # This currently encompasses cache invalidation counters and is CRuby-specific.
+      #
+      # JRuby emulates some CRuby global cache statistics, but they are synthetic and don't
+      # provide actionable performance information in the same way CRuby does.
+      # @see https://github.com/jruby/jruby/issues/4384#issuecomment-267069314
+      #
+      # TruffleRuby does not have a global runtime cache invalidation cache.
+      # @see http://archive.today/2021.09.10-205702/https://medium.com/graalvm/precise-method-and-constant-invalidation-in-truffleruby-4dd56c6bac1a
+      module VMCache
+        module_function
+
+        # Global constant cache "generation" counter.
+        #
+        # Whenever a constant creation busts the global constant cache
+        # this value is incremented. This has a measurable performance impact
+        # and thus show be avoided after application warm up.
+        def global_constant_state
+          ::RubyVM.stat[:global_constant_state]
+        end
+
+        # Global method cache "generation" counter.
+        #
+        # Whenever a method creation busts the global method cache
+        # this value is incremented. This has a measurable performance impact
+        # and thus show be avoided after application warm up.
+        #
+        # Since Ruby 3.0, the method class is kept on a per-class basis,
+        # largely mitigating global method cache busting. `global_method_state`
+        # is thus not available since Ruby 3.0.
+        # @see https://bugs.ruby-lang.org/issues/16614
+        def global_method_state
+          ::RubyVM.stat[:global_method_state]
+        end
```

Comment:
Great docs/research! I'm curious on the omission of `class_serial`, did you look into it and find it less useful?

Reply:
`class_serial` doesn't have a directly measurable negative performance impact on the application: it's a value that normally keeps going up on new class creation. But small class creation is a normal part of any sufficiently complex framework.
Because the fact of creating a class is in itself not harmful, I wasn't able to find a strong argument to report it, as I believe it won't be actionable. The other two method caches statistics in the PR captures any case where dynamic class/method modifications actually cause concerning issues.

Reply:
> But small class creation is a normal part of any sufficiently complex framework.

During request handling? I would've thought that it would "flatten out" at some point.

Happy to go with your decision; we can always revisit it if there's customer interest for that value.